migrate bcc based system monitor to libbpf and cilium/ebpf

daemon1024 opened the pull request:
**Why?**
- BCC compiles programs at runtime, which has heavy CPU and memory resource usage.
- eliminate `libbcc` dependency on the host machine
- smaller main image size, since compilation is decoupled from the main container.
- Making KubeArmor binary portable provided no CGO dependency

Follow Up #399 

Fixes #38 

TODO
- [x] Test systemd on centos/rhel
- [x] Handle GKE Headers
- [x] Update Docs

daemon1024 committed:
bpf: migrate bcc based ebpf code to libbpf

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
monitor: migrate system monitor from bcc to cilium ebpf

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
ci: update tests for new system monitor

updated docker container to compile ebpf code at runtime

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
build: update build system to use init container

- Use init container for compiling BPF code

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
deploy: update object to include init container

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
update systemd packaging

- setup CI to test systemd
- add llvm toolchain dep instead of bcc
- eliminate stdarg dep from clang/gcc std include

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 commented on KubeArmor/BPF/system_monitor.c:
```diff
@@ -15,6 +15,12 @@
 #define asm_inline asm
 #endif
```
Despite adding this in accordance with https://github.com/iovisor/bcc/issues/2546

Still getting the same errors when using the default version of clang provided in the `ubuntu-18.04` image in gh action, 
![image](https://user-images.githubusercontent.com/47106543/167141153-bd547fe9-6695-466f-9173-1f63884e54d5.png)
Ref https://github.com/kubearmor/KubeArmor/runs/6305329041?check_suite_focus=true

Any thoughts on why would this be happening?

It resolves if we upgrade the version of the clang which has the asm inline construct available, but still..

nam-jaehyun replied:
what clang version does Ubuntu-18.04 use? what is the minimum version of clang to make KubeArmor work?

daemon1024 replied:
> what clang version does Ubuntu-18.04 use? 

clang 9 in the GH Action env
clang 6 in the Vagrant env

> what is the minimum version of clang to make KubeArmor work?

It's not that straight forward
clang 6 works fine with kernel version 4.x.x
But with 5.x.x we need atleast clang 12

daemon1024 committed:
Replace BCC installation with latest llvm-ttolchain in setup script

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nam-jaehyun commented on Dockerfile:
```diff
@@ -1,12 +1,26 @@
+
+COPY ./KubeArmor/BPF /KubeArmor/BPF/
+COPY ./KubeArmor/build/compile.sh /KubeArmor/compile.sh
+
```
remove one line here

nam-jaehyun commented on KubeArmor/build/compile.sh:
```diff
@@ -0,0 +1,7 @@
+
+cd /KubeArmor/BPF
+make
+cp *.bpf.o /opt/kubearmor/BPF/
```
add a new line

nam-jaehyun commented on KubeArmor/build/entrypoint.sh:
```diff
@@ -15,23 +15,20 @@ DEBUG=0
+    ARMOR_OPTIONS=${@:2}
+    ;;
+*) ;;
+
```
remove an empty line

daemon1024 committed:
fix ebpf compilation on RHEL

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nyrahul commented on KubeArmor/BPF/include/stdarg.h:
```diff
@@ -0,0 +1,52 @@
+/*===---- stdarg.h - Variable argument handling ----------------------------===
```
duplicating standard headers! Why was this necessary?

daemon1024 replied:
I am really not sure now why we needed it 🙃 Removed it

nyrahul commented on KubeArmor/BPF/system_monitor.c:
```diff
@@ -1089,67 +1220,80 @@ TRACEPOINT_PROBE(syscalls, sys_exit_openat)
     return trace_ret_generic(_SYS_LISTEN, ctx, ARG_TYPE0(INT_T)|ARG_TYPE1(INT_T));
 }
+
+char LICENSE[] SEC("license") = "GPL";
```
```suggestion
char LICENSE[] SEC("license") = "Dual BSD/GPL";
```

nyrahul commented on KubeArmor/build/kubearmor-test-containerd.yaml:
```diff
@@ -37,13 +37,50 @@ spec:
+          - mountPath: /media/root/etc/os-release
+            name: os-release-path
+            readOnly: true
+          - mountPath: /usr/src
```
we will face problems with such `mountPath` .. gke COS latest rapid releases do not have `/usr/src` folder.
Ref: #648 

Is this yaml only for test purpose?

daemon1024 replied:
These YAML are only for test purposes. We will have to modify `deployments/get` with the new stuff for end user deployments and kArmor.

Adding that to TODO.

nyrahul commented on KubeArmor/monitor/systemMonitor.go:
```diff
@@ -169,7 +175,7 @@ func NewSystemMonitor(node *tp.Node, logger *fd.Feeder, containers *map[string]t
 	mon.Status = true
 	mon.UptimeTimeStamp = kl.GetUptimeTimestamp()
-	mon.HostByteOrder = bcc.GetHostByteOrder()
+	mon.HostByteOrder = binary.LittleEndian
```
can we not hardcode the endianness?

daemon1024 replied:
BCC used unsafe calls to determine the byte order
Ref https://github.com/iovisor/gobpf/blob/2289761f1e2092a7416cd5fd55a218802d997cb6/bcc/perf.go#L110

I would like to defer adding this till #614 and prefer using a better way like [conditional builds](https://groups.google.com/g/golang-nuts/c/3GEzwKfRRQw) to determine this. WDYT?

nyrahul replied:
sounds good. I believe ARM wont work anyways unless this gets fixed.

nyrahul reviewed (changes requested):
Brilliant piece of work!

Few comments other than specified inline.
I believe we will have issues with the mountPaths that are newly added ... 
Can we get this tested on GKE COS Regular/Rapid release channel images and on AWS Bottlerocket as well?

daemon1024 commented on KubeArmor/BPF/Makefile:
```diff
@@ -0,0 +1,86 @@
+	KRNDIR = /usr/src/$(KRNVER)
+else
+	KRNDIR = /usr/src/linux-headers-$(KRNVER)
+endif
```
Noting will need to handle GKE usr path here as well, because we are using the Makefile to compile in container.

daemon1024 replied:
Handled in entrypoint for init container (`compile.sh`)

daemon1024 committed:
chore: handle gke cos headers in init container

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
build: cleanup object files and update kernel header detection in init container

- Using lib modules for kernel header detection which symlinks to appropriate directory in usr src
- clean up object files and added object files to docker ignore

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nyrahul reviewed (approved):
LGTM 👍🏼 👍🏼

DelusionalOptimist commented on .gitignore:
```diff
@@ -17,6 +17,9 @@ contribution/vagrant/*.log
 .coverprofile
 
+#Object files
+*.o
```
@daemon1024 We should also remove object files present [here](https://github.com/kubearmor/KubeArmor/tree/main/KubeArmor/enforcer/bpflsm) from the tree, right?

daemon1024 replied:
Those are needed though, since they cannot be generated at runtime in GH Action (due to lack of BPFLSM support) . Thanks for noticing, I will update the ignore to be `BPF/*.o`

daemon1024 replied:
The object files in enforcer/bpflsm follow CO-RE ( Compile Once Run Everywhere ) hence it's fine for them to be in the tree, this is not the case for monitor code. Hence the need to clean them out.

DelusionalOptimist replied:
Gotcha :+1:

DelusionalOptimist reviewed (approved):
LGTM :rocket:
